### DISCUSSION TITLE
Custom Data Types

### DISCUSSION THREAD

**Discussion by JulesNP** [jacentino/DbFun#2] [Q&A] [answered]
First of all, thank you for creating this project! It's been great to work with so far.

I'm currently working with a Postgres database which uses [Xids](https://github.com/modfin/pg-xid) as primary keys.

I'm able to use the string data type for the Xids in my F# record when getting query results using DbFun, but if I need to extract the timestamp or other data from the Xid, I have to parse it each time from a string into an [Xid.Net](https://github.com/Yortw/Xid.Net) object.

What would be the best way for me to have DbFun automatically convert these Xid parameters to and from string types?

I found [an article about custom data type conversions](https://github.com/jacentino/SqlFun/wiki/Custom-conversions) on the SqlFun wiki, but it seems to be focusing strictly on collection types, and it's not clear how much of the example is translatable to the newer DbFun project.

**Comment by jacentino** [comment, marked as answer]
Hi Jules,

I'm happy you use my project. I'm quite proud of it and it's always good
news that people want to work with it.

Indeed, I haven't written about custom conversions in DbFun yet.
The config object has two methods for it:
* member this.AddParamConverter(convert: 'Source -> 'Target)
*  member this.AddRowConverter(convert: 'Source -> 'Target)

The first one registers function converting parameter values, the second
one converts column values to object fields.
Converters are just simple functions.

Jacek

czw., 7 lis 2024 o 18:37 JulesNP ***@***.***> napisał(a):

**Comment by JulesNP** [comment]
Thanks, that works great.
